File: project_euler/src/problem_054.erl
%% In the card game poker, a hand consists of five cards and are ranked, from lowest to highest, in the following way:
%%
%% 1) High Card: Highest value card.
%% 2) One Pair: Two cards of the same value.
%% 3) Two Pairs: Two different pairs.
%% 4) Three of a Kind: Three cards of the same value.
%% 5) Straight: All cards are consecutive values.
%% 6) Flush: All cards of the same suit.
%% 7) Full House: Three of a kind and a pair.
%% 8) Four of a Kind: Four cards of the same value.
%% 9) Straight Flush: All cards are consecutive values of same suit.
%% 10) Royal Flush: Ten, Jack, Queen, King, Ace, in same suit.
%%
%% The cards are valued in the order: 2, 3, 4, 5, 6, 7, 8, 9, 10, Jack, Queen, King, Ace.
%% If two players have the same ranked hands then the rank made up of the highest value wins; for example, a pair of eights beats a pair of fives (see example 1 below).
%% But if two ranks tie, for example, both players have a pair of queens, then highest cards in each hand are compared (see example 4 below);
%% if the highest cards tie then the next highest cards are compared, and so on.
%%
%% Consider the following five hands dealt to two players:
%% Hand : Player 1 : Player 2 : Winner
%% 1 : 5H 5C 6S 7S KD (Pair of Fives) : 2C 3S 8S 8D TD (Pair of Eights) : Player 2
%% 2 : 5D 8C 9S JS AC (Highest card Ace) : 2C 5C 7D 8S QH (Highest card Queen) : Player 1
%% 3 : 2D 9C AS AH AC (Three Aces) : 3D 6D 7D TD QD (Flush with Diamonds) : Player 2
%% 4 : 4D 6S 9H QH QC (Pair of Queens, Highest card Nine) : 3D 6D 7H QD QS (Pair of Queens, Highest card Seven) : Player 1
%% 5 : 2H 2D 4C 4D 4S (Full House with Three Fours) : 3C 3D 3S 9S 9D (Full House with Three Threes) : Player 1
%%
%% The file, problem_054.dat, contains one-thousand random hands dealt to two players.
%% Each line of the file contains ten cards (separated by a single space): the first five are Player 1's cards and the last five are Player 2's cards.
%% You can assume that all hands are valid (no invalid characters or repeated cards), each player's hand is in no specific order, and in each hand there is a clear winner.
%% How many hands does Player 1 win?

-module(problem_054).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

-define(TEN_CHAR, $T).
-define(JACK_CHAR, $J).
-define(QUEEN_CHAR, $Q).
-define(KING_CHAR, $K).
-define(ACE_CHAR, $A).

-define(TEN, 10).
-define(JACK, 11).
-define(QUEEN, 12).
-define(KING, 13).
-define(ACE, 14).

-define(HIGH_CARD, 1).
-define(ONE_PAIR, 2).
-define(TWO_PAIRS, 3).
-define(THREE_KIND, 4).
-define(STRAIGHT, 5).
-define(FLUSH, 6).
-define(FULL_HOUSE, 7).
-define(FOUR_KIND, 8).
-define(STRAIGHT_FLUSH, 9).
-define(ROYAL_FLUSH, 10).

get_check_data() ->
    [{"problem_054.dat", 376}].

prepare_data(ModuleSourceDir, Filename) ->
    load_utils:read_strings(filename:join(ModuleSourceDir, Filename)).

solve(HandsTable) ->
    CollectFun = fun(HandsRow, Count) ->
        Hands = prepare_hands(HandsRow),
        case check_hands(Hands) of
            first -> Count + 1;
            _Other -> Count
        end
    end,
    lists:foldl(CollectFun, 0, HandsTable).

-spec prepare_hands(HandsSource :: string()) -> {Hand1 :: [{Value :: pos_integer(), Suit :: char()}], Hand2 :: [{Value :: pos_integer(), Suit :: char()}]}.
prepare_hands(HandsSource) ->
    HandsData = string:tokens(HandsSource, " "),
    {Hand1, Hand2} = parse_hands(HandsData),
    {sort_cards(Hand1), sort_cards(Hand2)}.

-spec parse_hands(HandsData :: [string()]) -> {Hand1 :: [{Value :: pos_integer(), Suit :: char()}], Hand2 :: [{Value :: pos_integer(), Suit :: char()}]}.
parse_hands([Player1Card1, Player1Card2, Player1Card3, Player1Card4, Player1Card5, Player2Card1, Player2Card2, Player2Card3, Player2Card4, Player2Card5]) ->
    {[parse_card(Player1Card1), parse_card(Player1Card2), parse_card(Player1Card3),parse_card(Player1Card4), parse_card(Player1Card5)],
     [parse_card(Player2Card1), parse_card(Player2Card2), parse_card(Player2Card3),parse_card(Player2Card4), parse_card(Player2Card5)]}.

-spec parse_card(Card :: string()) -> {Value :: pos_integer(), Suit :: char()}.
parse_card([?TEN_CHAR, Suit]) -> {?TEN, Suit};
parse_card([?JACK_CHAR, Suit]) -> {?JACK, Suit};
parse_card([?QUEEN_CHAR, Suit]) -> {?QUEEN, Suit};
parse_card([?KING_CHAR, Suit]) -> {?KING, Suit};
parse_card([?ACE_CHAR, Suit]) -> {?ACE, Suit};
parse_card([Value, Suit]) -> {Value - $0, Suit}.

-spec sort_cards(Hand :: [{Value :: pos_integer(), Suit :: char()}]) -> [{Value :: pos_integer(), Suit :: char()}].
sort_cards(Hand) ->
    %% sorting in back order
    lists:sort(fun({Value1, _Suit1}, {Value2, _Suit2}) -> Value1 > Value2 end, Hand).

-spec calc_hand_rank([{Value :: pos_integer(), Suit :: char()}]) -> {Rank :: 1..10, MaxValue :: non_neg_integer(), MinValue :: non_neg_integer()}.
%% ROYAL FLUSH
calc_hand_rank([{?ACE, Suit}, {?KING, Suit}, {?QUEEN, Suit}, {?JACK, Suit}, {?TEN, Suit}]) -> {?ROYAL_FLUSH, 0, 0};
%% STRAIGHT FLUSH
calc_hand_rank([{Value1, Suit}, {Value2, Suit}, {Value3, Suit}, {Value4, Suit}, {Value5, Suit}])
    when Value2 == Value1 - 1, Value3 == Value1 - 2, Value4 == Value1 - 3, Value5 == Value1 - 4 -> {?STRAIGHT_FLUSH, Value1, Value5};
calc_hand_rank([{Value1, Suit}, {Value2, Suit}, {Value3, Suit}, {Value4, Suit}, {Value5, Suit}])
    when Value4 == Value5 + 1, Value3 == Value5 + 2, Value2 == Value5 + 3, Value1 == Value5 + 4 -> {?STRAIGHT_FLUSH, Value1, Value5};
%% FOUR OF A KIND
calc_hand_rank([{Value, _Suit1}, {Value, _Suit2}, {Value, _Suit3}, {Value, _Suit4}, {_Value5, _Suit5}]) -> {?FOUR_KIND, Value, 0};
calc_hand_rank([{_Value1, _Suit1}, {Value, _Suit2}, {Value, _Suit3}, {Value, _Suit4}, {Value, _Suit5}]) -> {?FOUR_KIND, Value, 0};
%% FULL HOUSE
calc_hand_rank([{Value1, _Suit1}, {Value1, _Suit2}, {Value3, _Suit3}, {Value3, _Suit4}, {Value3, _Suit5}]) -> {?FULL_HOUSE, Value1, Value3};
calc_hand_rank([{Value1, _Suit1}, {Value1, _Suit2}, {Value1, _Suit3}, {Value3, _Suit4}, {Value3, _Suit5}]) -> {?FULL_HOUSE, Value1, Value3};
%% FLUSH
calc_hand_rank([{_Value1, Suit}, {_Value2, Suit}, {_Value3, Suit}, {_Value4, Suit}, {_Value5, Suit}]) -> {?FLUSH, 0, 0};
%% STRAIGHT
calc_hand_rank([{Value1, _Suit1}, {Value2, _Suit2}, {Value3, _Suit3}, {Value4, _Suit4}, {Value5, _Suit5}])
    when Value2 == Value1 - 1, Value3 == Value1 - 2, Value4 == Value1 - 3, Value5 == Value1 - 4 -> {?STRAIGHT, Value1, Value5};
calc_hand_rank([{Value1, _Suit1}, {Value2, _Suit2}, {Value3, _Suit3}, {Value4, _Suit4}, {Value5, _Suit5}])
    when Value4 == Value5 + 1, Value3 == Value5 + 2, Value2 == Value5 + 3, Value1 == Value5 + 4 -> {?STRAIGHT, Value1, Value5};
%% THREE OF A KIND
calc_hand_rank([{Value, _Suit1}, {Value, _Suit2}, {Value, _Suit3}, {_Value4, _Suit4}, {_Value5, _Suit5}]) -> {?THREE_KIND, Value, 0};
calc_hand_rank([{_Value1, _Suit1}, {Value, _Suit2}, {Value, _Suit3}, {Value, _Suit4}, {_Value5, _Suit5}]) -> {?THREE_KIND, Value, 0};
calc_hand_rank([{_Value1, _Suit1}, {_Value2, _Suit2}, {Value, _Suit3}, {Value, _Suit4}, {Value, _Suit5}]) -> {?THREE_KIND, Value, 0};
%% TWO PAIRS
calc_hand_rank([{Value1, _Suit1}, {Value1, _Suit2}, {Value3, _Suit3}, {Value3, _Suit4}, {_Value5, _Suit5}]) -> {?TWO_PAIRS, Value1, Value3};
calc_hand_rank([{Value1, _Suit1}, {Value1, _Suit2}, {_Value3, _Suit3}, {Value4, _Suit4}, {Value4, _Suit5}]) -> {?TWO_PAIRS, Value1, Value4};
calc_hand_rank([{_Value1, _Suit1}, {Value2, _Suit2}, {Value2, _Suit3}, {Value4, _Suit4}, {Value4, _Suit5}]) -> {?TWO_PAIRS, Value2, Value4};
%% ONE PAIR
calc_hand_rank([{Value, _Suit1}, {Value, _Suit2}, {_Value3, _Suit3}, {_Value4, _Suit4}, {_Value5, _Suit5}]) -> {?ONE_PAIR, Value, 0};
calc_hand_rank([{_Value1, _Suit1}, {Value, _Suit2}, {Value, _Suit3}, {_Value4, _Suit4}, {_Value5, _Suit5}]) -> {?ONE_PAIR, Value, 0};
calc_hand_rank([{_Value1, _Suit1}, {_Value2, _Suit2}, {Value, _Suit3}, {Value, _Suit4}, {_Value5, _Suit5}]) -> {?ONE_PAIR, Value, 0};
calc_hand_rank([{_Value1, _Suit1}, {_Value2, _Suit2}, {_Value3, _Suit3}, {Value, _Suit4}, {Value, _Suit5}]) -> {?ONE_PAIR, Value, 0};
%% HIGH CARD
calc_hand_rank([_Card1, _Card2, _Card3, _Card4, _Card5]) -> {?HIGH_CARD, 0, 0}.

-spec check_hands({Hand1 :: [{Value :: pos_integer(), Suit :: char()}], Hand2 :: [{Value :: pos_integer(), Suit :: char()}]}) -> 'undefined' | 'first' | 'second'.
check_hands({Hand1, Hand2}) ->
    Rank1 = calc_hand_rank(Hand1),
    Rank2 = calc_hand_rank(Hand2),
    if
        Rank1 > Rank2 -> first;
        Rank1 < Rank2 -> second;
        Rank1 == Rank2 -> check_hands_cards(Hand1, Hand2)
    end.

-spec check_hands_cards(Card1 :: {Value :: pos_integer(), Suit :: char()}, Card2 :: {Value :: pos_integer(), Suit :: char()}) -> 'undefined' | 'first' | 'second'.
check_hands_cards([], []) -> undefined;
check_hands_cards([{Value1, _Suit1} | _Rest1], [{Value2, _Suit2} | _Rest2]) when Value1 > Value2 -> first;
check_hands_cards([{Value1, _Suit1} | _Rest1], [{Value2, _Suit2} | _Rest2]) when Value1 < Value2 -> second;
check_hands_cards([{Value, _Suit1} | Rest1], [{Value, _Suit2} | Rest2]) -> check_hands_cards(Rest1, Rest2).

File: projecteuler.fsharp/ProjectEulerTasks/Problem054.fs
﻿namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open System.IO
open System

// In the card game poker, a hand consists of five cards and are ranked, from lowest to highest, in the following way:
//
// 1) High Card: Highest value card.
// 2) One Pair: Two cards of the same value.
// 3) Two Pairs: Two different pairs.
// 4) Three of a Kind: Three cards of the same value.
// 5) Straight: All cards are consecutive values.
// 6) Flush: All cards of the same suit.
// 7) Full House: Three of a kind and a pair.
// 8) Four of a Kind: Four cards of the same value.
// 9) Straight Flush: All cards are consecutive values of same suit.
// 10) Royal Flush: Ten, Jack, Queen, King, Ace, in same suit.
//
// The cards are valued in the order: 2, 3, 4, 5, 6, 7, 8, 9, 10, Jack, Queen, King, Ace.
// If two players have the same ranked hands then the rank made up of the highest value wins; for example, a pair of eights beats a pair of fives (see example 1 below).
// But if two ranks tie, for example, both players have a pair of queens, then highest cards in each hand are compared (see example 4 below); if the highest cards tie then the next highest cards are compared, and so on.
//
// Consider the following five hands dealt to two players:
// Hand : Player 1 : Player 2 : Winner
// 1 : 5H 5C 6S 7S KD (Pair of Fives) : 2C 3S 8S 8D TD (Pair of Eights) : Player 2
// 2 : 5D 8C 9S JS AC (Highest card Ace) : 2C 5C 7D 8S QH (Highest card Queen) : Player 1
// 3 : 2D 9C AS AH AC (Three Aces) : 3D 6D 7D TD QD (Flush with Diamonds) : Player 2
// 4 : 4D 6S 9H QH QC (Pair of Queens, Highest card Nine) : 3D 6D 7H QD QS (Pair of Queens, Highest card Seven) : Player 1
// 5 : 2H 2D 4C 4D 4S (Full House with Three Fours) : 3C 3D 3S 9S 9D (Full House with Three Threes) : Player 1
//
// The file, problem_054.dat, contains one-thousand random hands dealt to two players.
// Each line of the file contains ten cards (separated by a single space): the first five are Player 1's cards and the last five are Player 2's cards.
// You can assume that all hands are valid (no invalid characters or repeated cards), each player's hand is in no specific order, and in each hand there is a clear winner.
// How many hands does Player 1 win?

module Problem054Impl =

    [<Literal>]
    let JackValue = 11

    [<Literal>]
    let QueenValue = 12

    [<Literal>]
    let KingValue = 13

    [<Literal>]
    let AceValue = 14

    [<Literal>]
    let HighCardRank = 1

    [<Literal>]
    let OnePairRank = 2

    [<Literal>]
    let TwoPairsRank = 3

    [<Literal>]
    let ThreeKindRank = 4

    [<Literal>]
    let StraightRank = 5

    [<Literal>]
    let FlushRank = 6

    [<Literal>]
    let FullHouseRank = 7

    [<Literal>]
    let FourKindRank = 8

    [<Literal>]
    let StraightFlushRank = 9

    [<Literal>]
    let RoyalFlushRank = 10

    type Card = {Value: int; Suit: char}

    type Hands = {Player1: List<Card>; Player2: List<Card>}

    type HandRank =
    | HighCard of card: Card
    | OnePair of card: Card
    | TwoPairs of card1: Card * card2: Card
    | ThreeKind of card: Card
    | Straight of highCard: Card
    | Flush
    | FullHouse of card1: Card * card2: Card
    | FourKind of card: Card
    | StraightFlush of highCard: Card
    | RoyalFlush

        member public this.GetValue() =
            match this with
            | HighCard _ -> HighCardRank
            | OnePair _ -> OnePairRank
            | TwoPairs _ -> TwoPairsRank
            | ThreeKind _ -> ThreeKindRank
            | Straight _ -> StraightRank
            | Flush -> FlushRank
            | FullHouse _ -> FullHouseRank
            | FourKind _ -> FourKindRank
            | StraightFlush _ -> StraightFlushRank
            | RoyalFlush -> RoyalFlushRank

    type HandsWinner =
    | Player1
    | Player2
    | Draw

open Problem054Impl

[<TestFixture>]
type Problem054() =

    let playerCardsCount = 5

    let parseHands (handsData: string) =
        let parseCard = function
            | 'A' :: suit :: [] -> {Card.Value = AceValue; Card.Suit = suit}
            | 'K' :: suit :: [] -> {Card.Value = KingValue; Card.Suit = suit}
            | 'Q' :: suit :: [] -> {Card.Value = QueenValue; Card.Suit = suit}
            | 'J' :: suit :: [] -> {Card.Value = JackValue; Card.Suit = suit}
            | 'T' :: suit :: [] -> {Card.Value = 10; Card.Suit = suit}
            | value :: suit :: [] -> {Card.Value = Char.GetNumericValue(value) |> int; Card.Suit = suit}
            | _ -> failwith "Unknown format of data"
        let cardsData = handsData.Split(' ')
        let player1Cards = cardsData |> Seq.take playerCardsCount |> Seq.map (fun cardData -> cardData |> Seq.toList |> parseCard) |> Seq.sortByDescending (fun card -> card.Value) |> Seq.toList
        let player2Cards = cardsData |> Seq.skip playerCardsCount |> Seq.map (fun cardData -> cardData |> Seq.toList |> parseCard) |> Seq.sortByDescending (fun card -> card.Value) |> Seq.toList
        {Hands.Player1 = player1Cards; Hands.Player2 = player2Cards}

    let (|RoyalFlush|_|) = function
        | [{Card.Value = AceValue; Card.Suit = suit1};
           {Card.Value = KingValue; Card.Suit = suit2};
           {Card.Value = QueenValue; Card.Suit = suit3};
           {Card.Value = JackValue; Card.Suit = suit4};
           {Card.Value = 10; Card.Suit = suit5}] when (suit1 = suit2) && (suit2 = suit3) && (suit3 = suit4) && (suit4 = suit5) -> Some ()
        | _ -> None

    let (|StraightFlush|_|) = function
        | [{Card.Value = value1; Card.Suit = suit1};
           {Card.Value = value2; Card.Suit = suit2};
           {Card.Value = value3; Card.Suit = suit3};
           {Card.Value = value4; Card.Suit = suit4};
           {Card.Value = value5; Card.Suit = suit5}]
          when (suit1 = suit2) && (suit2 = suit3) && (suit3 = suit4) && (suit4 = suit5) && (value1 - value2 = 1) && (value2 - value3 = 1) && (value3 - value4 = 1) && (value4 - value5 = 1) -> Some {Card.Value = value1; Card.Suit = suit1}
        | _ -> None

    let (|FourKind|_|) = function
        | [{Card.Value = value1; Card.Suit = suit}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 = value3) && (value3 = value4) && (value4 <> value5) -> Some {Card.Value = value1; Card.Suit = suit}
        | [{Card.Value = value1}; {Card.Value = value2; Card.Suit = suit}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 = value3) && (value3 = value4) && (value4 = value5) -> Some {Card.Value = value2; Card.Suit = suit}
        | _ -> None

    let (|FullHouse|_|) = function
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4; Card.Suit = suit4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 = value3) && (value3 <> value4) && (value4 = value5) -> Some ({Card.Value = value1; Card.Suit = suit1}, {Card.Value = value4; Card.Suit = suit4})
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3; Card.Suit = suit3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 <> value3) && (value3 = value4) && (value4 = value5) -> Some ({Card.Value = value1; Card.Suit = suit1}, {Card.Value = value3; Card.Suit = suit3})
        | _ -> None

    let (|Flush|_|) = function
        | [{Card.Suit = suit1}; {Card.Suit = suit2}; {Card.Suit = suit3}; {Card.Suit = suit4}; {Card.Suit = suit5}]
          when (suit1 = suit2) && (suit2 = suit3) && (suit3 = suit4) && (suit4 = suit5) -> Some ()
        | _ ->  None

    let (|Straight|_|) = function
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 - value2 = 1) && (value2 - value3 = 1) && (value3 - value4 = 1) && (value4 - value5 = 1) -> Some {Card.Value = value1; Card.Suit = suit1}
        | _ -> None

    let (|ThreeKind|_|) = function
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 = value3) && (value3 <> value4) && (value4 <> value5) -> Some {Card.Value = value1; Card.Suit = suit1}
        | [{Card.Value = value1}; {Card.Value = value2; Card.Suit = suit2}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 = value3) && (value3 = value4) && (value4 <> value5) -> Some {Card.Value = value2; Card.Suit = suit2}
        | [{Card.Value = value1}; {Card.Value = value2}; {Card.Value = value3; Card.Suit = suit3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 <> value3) && (value3 = value4) && (value4 = value5) -> Some {Card.Value = value3; Card.Suit = suit3}
        | _ -> None

    let (|TwoPairs|_|) = function
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3; Card.Suit = suit3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 <> value3) && (value3 = value4) && (value4 <> value5) -> Some ({Card.Value = value1; Card.Suit = suit1}, {Card.Value = value3; Card.Suit = suit3})
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4; Card.Suit = suit4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 <> value3) && (value3 <> value4) && (value4 = value5) -> Some ({Card.Value = value1; Card.Suit = suit1}, {Card.Value = value4; Card.Suit = suit4})
        | [{Card.Value = value1}; {Card.Value = value2; Card.Suit = suit2}; {Card.Value = value3}; {Card.Value = value4; Card.Suit = suit4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 = value3) && (value3 <> value4) && (value4 = value5) -> Some ({Card.Value = value2; Card.Suit = suit2}, {Card.Value = value4; Card.Suit = suit4})
        | _ -> None

    let (|OnePair|_|) = function
        | [{Card.Value = value1; Card.Suit = suit1}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 = value2) && (value2 <> value3) && (value3 <> value4) && (value4 <> value5) -> Some {Card.Value = value1; Card.Suit = suit1}
        | [{Card.Value = value1}; {Card.Value = value2; Card.Suit = suit2}; {Card.Value = value3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 = value3) && (value3 <> value4) && (value4 <> value5) -> Some {Card.Value = value2; Card.Suit = suit2}
        | [{Card.Value = value1}; {Card.Value = value2}; {Card.Value = value3; Card.Suit = suit3}; {Card.Value = value4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 <> value3) && (value3 = value4) && (value4 <> value5) -> Some {Card.Value = value3; Card.Suit = suit3}
        | [{Card.Value = value1}; {Card.Value = value2}; {Card.Value = value3}; {Card.Value = value4; Card.Suit = suit4}; {Card.Value = value5}]
          when (value1 <> value2) && (value2 <> value3) && (value3 <> value4) && (value4 = value5) -> Some {Card.Value = value4; Card.Suit = suit4}
        | _ -> None

    let calcHandRank = function
        | RoyalFlush -> HandRank.RoyalFlush
        | StraightFlush highCard -> HandRank.StraightFlush highCard
        | FourKind card -> HandRank.FourKind card
        | FullHouse (card1, card2) -> HandRank.FullHouse (card1, card2)
        | Flush -> HandRank.Flush
        | Straight highCard -> HandRank.Straight highCard
        | ThreeKind card -> HandRank.ThreeKind card
        | TwoPairs (card1, card2) -> HandRank.TwoPairs (card1, card2)
        | OnePair card -> HandRank.OnePair card
        | {Card.Value= value1; Card.Suit = suit1} :: _ -> HandRank.HighCard {Card.Value = value1; Card.Suit = suit1}
        | _ -> failwith "Unexpected branch of match expression"

    let rec compareCards (player1: List<Card>) (player2: List<Card>) =
        match player1, player2 with
        | [], [] -> HandsWinner.Draw
        | player1Card :: _, player2Card :: _ when player1Card.Value > player2Card.Value -> HandsWinner.Player1
        | player1Card :: _, player2Card :: _ when player1Card.Value < player2Card.Value -> HandsWinner.Player2
        | _ :: player1CardsRest, _ :: player2CardsRest -> compareCards player1CardsRest player2CardsRest
        | _ -> failwith "Unexpected branch of match expression"

    let compareHands {Hands.Player1 = player1Cards; Hands.Player2 = player2Cards} =
        let player1HandRank = player1Cards |> calcHandRank
        let player2HandRank = player2Cards |> calcHandRank
        match player1HandRank, player2HandRank with
        | _ when player1HandRank.GetValue() > player2HandRank.GetValue() -> HandsWinner.Player1
        | _ when player1HandRank.GetValue() < player2HandRank.GetValue() -> HandsWinner.Player2
        | HandRank.RoyalFlush, HandRank.RoyalFlush -> HandsWinner.Draw
        | HandRank.StraightFlush highCard1, HandRank.StraightFlush highCard2 when highCard1.Value > highCard2.Value -> HandsWinner.Player1
        | HandRank.StraightFlush highCard1, HandRank.StraightFlush highCard2 when highCard1.Value < highCard2.Value -> HandsWinner.Player2
        | HandRank.StraightFlush highCard1, HandRank.StraightFlush highCard2 when highCard1.Value = highCard2.Value -> HandsWinner.Draw
        | HandRank.FourKind card1, HandRank.FourKind card2 when card1.Value > card2.Value -> HandsWinner.Player1
        | HandRank.FourKind card1, HandRank.FourKind card2 when card1.Value < card2.Value -> HandsWinner.Player2
        | HandRank.FullHouse (threeCard1, _), HandRank.FullHouse (threeCard2, _) when threeCard1.Value > threeCard2.Value -> HandsWinner.Player1
        | HandRank.FullHouse (threeCard1, _), HandRank.FullHouse (threeCard2, _) when threeCard1.Value < threeCard2.Value -> HandsWinner.Player2
        | HandRank.FullHouse (threeCard1, pairCard1), HandRank.FullHouse (threeCard2, pairCard2) when (threeCard1.Value = threeCard2.Value) && (pairCard1.Value > pairCard2.Value) -> HandsWinner.Player1
        | HandRank.FullHouse (threeCard1, pairCard1), HandRank.FullHouse (threeCard2, pairCard2) when (threeCard1.Value = threeCard2.Value) && (pairCard1.Value < pairCard2.Value) -> HandsWinner.Player2
        | HandRank.FullHouse (threeCard1, pairCard1), HandRank.FullHouse (threeCard2, pairCard2) when (threeCard1.Value = threeCard2.Value) && (pairCard1.Value = pairCard2.Value) -> HandsWinner.Draw
        | HandRank.Straight highCard1, HandRank.Straight highCard2 when highCard1.Value > highCard2.Value -> HandsWinner.Player1
        | HandRank.Straight highCard1, HandRank.Straight highCard2 when highCard1.Value < highCard2.Value -> HandsWinner.Player2
        | HandRank.Straight highCard1, HandRank.Straight highCard2 when highCard1.Value = highCard2.Value -> HandsWinner.Draw
        | HandRank.ThreeKind card1, HandRank.ThreeKind card2 when card1.Value > card2.Value -> HandsWinner.Player1
        | HandRank.ThreeKind card1, HandRank.ThreeKind card2 when card1.Value < card2.Value -> HandsWinner.Player2
        | HandRank.TwoPairs (pair1Card1, _), HandRank.TwoPairs (pair1Card2, _) when pair1Card1.Value > pair1Card2.Value -> HandsWinner.Player1
        | HandRank.TwoPairs (pair1Card1, _), HandRank.TwoPairs (pair1Card2, _) when pair1Card1.Value < pair1Card2.Value -> HandsWinner.Player2
        | HandRank.TwoPairs (pair1Card1, pair2Card1), HandRank.TwoPairs (pair1Card2, pair2Card2) when (pair1Card1.Value = pair1Card2.Value) && (pair2Card1.Value > pair2Card2.Value) -> HandsWinner.Player1
        | HandRank.TwoPairs (pair1Card1, pair2Card1), HandRank.TwoPairs (pair1Card2, pair2Card2) when (pair1Card1.Value = pair1Card2.Value) && (pair2Card1.Value < pair2Card2.Value) -> HandsWinner.Player2
        | HandRank.OnePair card1, HandRank.OnePair card2 when card1.Value > card2.Value -> HandsWinner.Player1
        | HandRank.OnePair card1, HandRank.OnePair card2 when card1.Value < card2.Value -> HandsWinner.Player2
        | _ -> compareCards player1Cards player2Cards

    let solveImpl (dataFilename: string) =
        File.ReadAllLines(Path.Combine("Data", dataFilename)) |> Seq.map parseHands |> Seq.map compareHands |> Seq.filter (fun winner -> winner = HandsWinner.Player1) |> Seq.length

    [<TestCase("problem_054.dat", 376, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(dataFilename: string, expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, dataFilename)